src/tunnel.cc

```diff
@@ -1275,8 +1289,9 @@ TunnelStateData::noteDestinationsEnd(ErrorState *selectionError)
         if (selectionError)
             return sendError(selectionError, "path selection has failed");
 
+        // TODO: Merge with FwdState and remove this likely unnecessary check.
         if (savedError)
-            return sendError(savedError, "all found paths have failed");
+            return sendError(savedError, "path selection found no paths (with an impossible early error)");
```

Comment:
What does "impossible early error" mean here ?

If it is truly impossible for savedError to be non-nul this should be an assert instead of an if-statement

Reply:
>     * impossible: I believe that such early errors do not exist today, but I am not sure and consider figuring the correct answer to that existence question difficult out-of-scope (and possibly even unnecessary due to other XXXs) work.
> 

Given that, I suggest "very early" to avoid others like me confusing this with a certainty.

> 
> > If it is truly impossible for savedError to be non-nul this should be an assert instead of an if-statement
> 
> Agreed. If we had Assure(), I might have had the courage to put it in here in this PR, but we still do not have Assure(), and convincing myself that an assert() is safe here is difficult out-of-scope (and possibly unnecessary) work.

That would be why Must() currently gets used in places like this.

Reply:
```
"path selection found no paths (with an impossible early error)"
```

> > * early: Here, roughly speaking, it is an error that occurred before TunnelStateData received either the first destination or a destination search failure notification.
> > * impossible: I believe that such early errors do not exist today, but I am not sure and consider figuring the correct answer to that existence question difficult out-of-scope (and possibly even unnecessary due to other XXXs) work.
 
> Given that, I suggest "very early" to avoid others like me confusing this with a certainty.

I do not understand how replacing "early" with "very early" reduces _certainty_ for others like you. For me, it would only introduce a third class of errors -- the "very early" ones -- something I would rather not do. Would you prefer "a likely impossible early error" phrasing instead? Adding "likely" would indeed decrease certainty (for any reader) IMO.
 
> > > If it is truly impossible for savedError to be non-nul this should be an assert instead of an if-statement
 
> > Agreed. If we had Assure(), I might have had the courage to put it in here in this PR, but we still do not have Assure(), and convincing myself that an assert() is safe here is difficult out-of-scope (and possibly unnecessary) work.
 
> That would be why Must() currently gets used in places like this. 

Exactly.

> Would you dare join the rabble in using it?

Not in this PR. As I said, investigating whether those errors are indeed impossible requires serious out of scope (and likely wasteful) work. On the other hand, if this PR starts killing these transactions (because those early errors happen to be possible after all), that would be an out of scope change (and a bug).